ADOLC Mat Vec

antoinefalisse opened the pull request:
@sherm1, @chrisdembia, this PR contains the changes in `Mat.h `and `Vec.h` necessary to build Simbody with ADOL-C. This PR also contains various tests in `TestADOLCCommon.cpp` verifying that it works properly. EDIT: parent branches have been merged. This PR is thus the next one in the ADOL-C queue.

One question about the tests, should the following operations be supported:

- `a/v` where a is an `adouble `and v a `Vec`
- `a-v` where a is an `adouble `and v a `Vec`
- `a-m` where a is an `adouble `and m a `Mat`

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/simbody/simbody/614)
<!-- Reviewable:end -->

chrisdembia commented:
@antoinefalisse I reviewed these changes and they look good. Once we decide what to do about the 3 operators you list, it would be good to add relevant comments.

@sherm1 here are compiler errors for the 3 operations @antoinefalisse raised above (using doubles, on the master branch).

### Vec division

```cpp
         double s = 5;
         Vec3 v(1, 2, 3);
         Row3 result = s / v;
```

```cpp
/Users/chris/repos/simbody/2/simbody/SimTKcommon/SmallMatrix/include/SimTKcommon/internal/Vec.h:1179:55: error:
      no member named 'DvdOp' in 'SimTK::NTraits<double>::Result<SimTK::Vec<3, double, 1> >'
  { return CNT<double>::template Result<Vec<M,E,S> >::DvdOp::perform(l,r); }
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
/Users/chris/repos/simbody/2/simbody/SimTKcommon/tests/TestSmallMatrix.cpp:493:25: note: in
      instantiation of function template specialization 'SimTK::operator/<3, double, 1>'
      requested here
        Row3 result = s / v;
                        ^
```

### Vec subtraction

```cpp
         double s = 5;
         Vec3 v(1, 2, 3);
         Vec3 result = s - v;
```

```cpp
/Users/chris/repos/simbody/2/simbody/SimTKcommon/SmallMatrix/include/SimTKcommon/internal/Vec.h:1296:55: error:
      no member named 'SubOp' in 'SimTK::NTraits<double>::Result<SimTK::Vec<3, double, 1> >'
  { return CNT<double>::template Result<Vec<M,E,S> >::SubOp::perform(l,r); }
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
/Users/chris/repos/simbody/2/simbody/SimTKcommon/tests/TestSmallMatrix.cpp:492:25: note: in
      instantiation of function template specialization 'SimTK::operator-<3, double, 1>'
      requested here
        Vec3 result = s - v;
                        ^
```

### Mat subtraction

```cpp
         double s = 5;
         Mat33 m(0);
         Mat33 result = s - m;
```

```cpp
/Users/chris/repos/simbody/2/simbody/SimTKcommon/SmallMatrix/include/SimTKcommon/internal/Mat.h:1495:61: error:
      no member named 'SubOp' in
      'SimTK::NTraits<double>::Result<SimTK::Mat<3, 3, double, 3, 1> >'
  { return CNT<double>::template Result<Mat<M,N,E,CS,RS> >::SubOp::perform(l,r); }
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
/Users/chris/repos/simbody/2/simbody/SimTKcommon/tests/TestSmallMatrix.cpp:492:26: note: in
      instantiation of function template specialization 'SimTK::operator-<3, 3, double, 3, 1>'
      requested here
        Mat33 result = s - m;
                         ^
```

I get similar errors if using `auto return = ...;` instead of explicitly specifying the return type.

---

Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[SimTKcommon/tests/TestADOLCCommon.cpp, line 313 at r2](https://reviewable.io:443/reviews/simbody/simbody/614#-L4Dee17Mw4aVCBXGLtY:-L4Dee17Mw4aVCBXGLtZ:b10ca2z) ([raw file](https://github.com/simbody/simbody/blob/858b4e2f1f62070785631ef1fd84b2b832686fb3/SimTKcommon/tests/TestADOLCCommon.cpp#L313)):*
> ```cpp
>     SimTK_TEST(mressr[1][0] == m[1][0]);
>     SimTK_TEST(mressr[0][1] == m[0][1]);
>     SimTK_TEST(mressr[1][1] == e-a);
> ```

Nice tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/614#-:-L4DcqsEmgTRbRXV32my:bg5zikw)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
+@sherm1 <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
These look good. I don't see any reason to support the three odd operations you mentioned:
- a/v where a is an adouble and v a Vec
- a-v where a is an adouble and v a Vec
- a-m where a is an adouble and m a Mat

The other direction does make sense: v/a, v-a, m-a (in the last two cases the scalar is treated like a diagonal matrix). Also the a/m where m is a matrix makes sense and is treated as `a*m.inverse()`.

---

Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/614#-:-L56oSwd5Cv5u4Z33GdY:b-eniscs)*
<!-- Sent from Reviewable.io -->

antoinefalisse commented:
Great thanks for the review! I have added a couple of comments (mainly copied from the headers) to the tests and corrected typos (looks like I had a real problem with "subtraction"). @chrisdembia, were you thinking of other comments:

> it would be good to add relevant comments.

---

Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/614#-:-L59aBgu0_A1mAEe4g82:beomxhs)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
@antoinefalisse I would remove the operators for the 3 above and add comments in those spots saying "The operation a/v where a is an adouble and v is a Vec is not supported." (similar for the other 2 operations).

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[SimTKcommon/SmallMatrix/include/SimTKcommon/internal/Vec.h, line 1199 at r3](https://reviewable.io:443/reviews/simbody/simbody/614#-L5APXj8RFIFRIyfjtg2:-L5APXj8RFIFRIyfjtg3:b-bmpj3e) ([raw file](https://github.com/simbody/simbody/blob/c8a89a39d7c3fb1a15e010cf6ac120b304335be9/SimTKcommon/SmallMatrix/include/SimTKcommon/internal/Vec.h#L1199)):*
> ```objc
>     typename CNT<adouble>::template Result<Vec<M,E,S> >::Dvd
>     operator/(const adouble& l, const Vec<M,E,S>& r)
>     { return CNT<adouble>::template Result<Vec<M,E,S> >::DvdOp::perform(l,r); }
> ```

Since we decided to not support this operation, I would remove this operator (and the other operators we agreed to not support).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/614#-:-L5AP4XFnu-Re4tQhYkE:bw4zaa2)*
<!-- Sent from Reviewable.io -->

antoinefalisse commented:
Ok thanks. This is now done. I think this PR is now ready to be merged.

---

Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[SimTKcommon/SmallMatrix/include/SimTKcommon/internal/Vec.h, line 1199 at r3](https://reviewable.io:443/reviews/simbody/simbody/614#-L5APXj8RFIFRIyfjtg2:-L5AUq9tMMJ9eQVzZKVi:b-896fix) ([raw file](https://github.com/simbody/simbody/blob/c8a89a39d7c3fb1a15e010cf6ac120b304335be9/SimTKcommon/SmallMatrix/include/SimTKcommon/internal/Vec.h#L1199)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary><blockquote>

Since we decided to not support this operation, I would remove this operator (and the other operators we agreed to not support).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/614#-:-L5AUqzkiMkEbg1qhalY:bdse456)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/614#-:-L5AVTMpbycuSH76sMq6:bnfp4nl)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/614#-:-L5Lu0WaDPo4LD8Bp2FF:bnfp4nl)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[SimTKcommon/tests/TestADOLCCommon.cpp, line 313 at r2](https://reviewable.io:443/reviews/simbody/simbody/614#-L4Dee17Mw4aVCBXGLtY:-L5LuEUjO9T0jWRY2T2H:b32w0x) ([raw file](https://github.com/simbody/simbody/blob/858b4e2f1f62070785631ef1fd84b2b832686fb3/SimTKcommon/tests/TestADOLCCommon.cpp#L313)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary><blockquote>

Nice tests.
</blockquote></details>

OK

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/614)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
@antoinefalisse could you resolve the remaining "discussion" on Reviewable? After that I think we can merge this.

antoinefalisse commented:
Done

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[SimTKcommon/tests/TestADOLCCommon.cpp, line 313 at r2](https://reviewable.io:443/reviews/simbody/simbody/614#-L4Dee17Mw4aVCBXGLtY:-L5LyUvQV6kQ5ESoZT7-:b-896fix) ([raw file](https://github.com/simbody/simbody/blob/858b4e2f1f62070785631ef1fd84b2b832686fb3/SimTKcommon/tests/TestADOLCCommon.cpp#L313)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary><blockquote>

OK
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/614#-:-L5LyYNl4Cnzv6XfT3R4:bb6un1j)*
<!-- Sent from Reviewable.io -->
